pkg/agent/nodeportlocal/k8s/controller.go

```diff
@@ -19,43 +19,95 @@ package k8s
 import (
 	"encoding/json"
 	"fmt"
+	"reflect"
 	"sync"
 	"time"
 
 	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/portcache"
 
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/apimachinery/pkg/util/wait"
 	clientset "k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
 	"k8s.io/client-go/tools/cache"
 	"k8s.io/client-go/util/workqueue"
 	"k8s.io/klog"
 )
 
 const (
-	minRetryDelay = 2 * time.Second
-	maxRetryDelay = 120 * time.Second
-	numWorkers    = 4
+	controllerName = "AntreaAgentNPLController"
+	minRetryDelay  = 2 * time.Second
+	maxRetryDelay  = 120 * time.Second
+	numWorkers     = 4
 )
 
-type Controller struct {
-	portTable  *portcache.PortTable
-	kubeClient clientset.Interface
-	queue      workqueue.RateLimitingInterface
-	Ctrl       cache.Controller
-	CacheStore cache.Store
-	PodToIP    map[string]string
-	podIPLock  sync.RWMutex
+type NPLController struct {
+	portTable   *portcache.PortTable
+	kubeClient  clientset.Interface
+	queue       workqueue.RateLimitingInterface
+	podInformer cache.SharedIndexInformer
+	podLister   corelisters.PodLister
+	svcInformer cache.SharedIndexInformer
+	podToIP     map[string]string
+	podIPLock   sync.RWMutex
 }
 
-func NewNPLController(kubeClient clientset.Interface, pt *portcache.PortTable) *Controller {
-	nplCtrl := Controller{
-		kubeClient: kubeClient,
-		portTable:  pt,
-		PodToIP:    make(map[string]string),
+func NewNPLController(kubeClient clientset.Interface,
+	podInformer cache.SharedIndexInformer,
+	svcInformer cache.SharedIndexInformer,
+	resyncPeriod time.Duration,
+	pt *portcache.PortTable) *NPLController {
+	c := NPLController{
+		kubeClient:  kubeClient,
+		portTable:   pt,
+		podInformer: podInformer,
+		podLister:   corelisters.NewPodLister(podInformer.GetIndexer()),
+		svcInformer: svcInformer,
+		podToIP:     make(map[string]string),
 	}
-	nplCtrl.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "nodeportlocal")
-	return &nplCtrl
+
+	podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    c.enqueuePod,
+			DeleteFunc: c.enqueuePod,
+			UpdateFunc: func(old, cur interface{}) {
```

Comment:
Does it make sense to check if Pod labels changed or not? @tnqn

Reply:
Changes to `ContainerPort` will also need to be monitored, as this eventually determines the iptable rule. Same goes for `podIP` as well but i'm not sure in what scenarios would the `podIP` change.
Also changing a `namespace` will affect the cache as the key would change.

We also put an annotation (`nodeportlocal.antrea.io: {nodeIP, nodePort, podPort}`) on the pod after successfully programming the iptable rule, so if someone changes/removes that, and we are not watching for annotation updates, then we won't be able to fix/add back our annotation.

I guess we can keep this one as it is for now.

Reply:
Sure. We can keep the current version for now. But note none of containerPort, podIP, annotation changes are common, and the Pod event processing is not simple, if we see multiple Pod events are typically generated and cause unnecessary computation, probably we should still consider optimization.

Here can you add some comments to explain what Pod changes (containerPort, IP, labels, NPL annotation) we are interested in? Namespace cannot be changed after a resource is created.

Reply:
> Changes to ContainerPort will also need to be monitored, as this eventually determines the iptable rule. Same goes for podIP as well but i'm not sure in what scenarios would the podIP change.
Also changing a namespace will affect the cache as the key would change.

Most of fields in spec are actually immutable. 
```
spec: Forbidden: pod updates may not change fields other than `spec.containers[*].image`, `spec.initContainers[*].image`, `spec.activeDeadlineSeconds` or `spec.tolerations`
```
So ContainerPort cannot change. `PodIP` may change if Sandbox container is dead as kubelet will recreate it and call CNI again.